platform/packages/finance/src/price/base/mod.rs

```diff
@@ -0,0 +1,251 @@
+)]
+pub struct BasePrice<BaseG, QuoteC, QuoteG>
+where
+    BaseG: Group,
+    QuoteC: Currency,
+    QuoteG: Group,
+{
+    amount: CoinDTO<BaseG>,
+    #[serde(serialize_with = "serialize_amount_quote::<_, _, QuoteG>")]
+    amount_quote: Coin<QuoteC>,
+    #[serde(skip)]
+    quote_group: PhantomData<QuoteG>,
+}
+
+impl<BaseG, QuoteC, QuoteG> BasePrice<BaseG, QuoteC, QuoteG>
+where
+    BaseG: Group,
+    QuoteC: Currency,
+    QuoteG: Group,
+{
+    fn new_checked(amount: CoinDTO<BaseG>, amount_quote: Coin<QuoteC>) -> FinanceResult<Self> {
+        let res = Self::new_raw(amount, amount_quote);
+        res.invariant_held().map(|_| res)
+    }
+
+    fn new_unchecked(amount: CoinDTO<BaseG>, amount_quote: Coin<QuoteC>) -> Self {
+        let res = Self::new_raw(amount, amount_quote);
+
+        debug_assert_eq!(Ok(()), res.invariant_held());
+        res
+    }
+
+    pub fn base_ticker(&self) -> &SymbolSlice {
+        self.amount.ticker()
+    }
+
+    pub(crate) fn amount(&self) -> &CoinDTO<BaseG> {
+        &self.amount
+    }
+
+    pub(crate) fn amount_quote(&self) -> Coin<QuoteC> {
+        self.amount_quote
+    }
+
+    fn new_raw(amount: CoinDTO<BaseG>, amount_quote: Coin<QuoteC>) -> Self {
+        Self {
+            amount,
+            amount_quote,
+            quote_group: PhantomData,
+        }
+    }
+
+    fn invariant_held(&self) -> FinanceResult<()> {
+        Self::check(!self.amount.is_zero(), "The amount should not be zero").and_then(|_| {
+            Self::check(
+                !self.amount_quote.is_zero(),
+                "The quote amount should not be zero",
+            )
+        })
+    }
+
+    fn check(invariant: bool, msg: &str) -> FinanceResult<()> {
+        Error::broken_invariant_if::<Self>(!invariant, msg)
+    }
+}
+
+fn serialize_amount_quote<S, QuoteC, QuoteG>(
+    amount: &Coin<QuoteC>,
+    serializer: S,
+) -> Result<S::Ok, S::Error>
+where
+    S: serde::Serializer,
+    QuoteC: Currency + ?Sized,
+    QuoteG: Group,
+{
+    currency::validate_member::<QuoteC, QuoteG>()
+        .map_err(|err| {
+            serde::ser::Error::custom(format!("Amount quote serializaion failed: {:?}", err))
+        })
+        .and_then(|_| {
+            let coin_dto = CoinDTO::<QuoteG>::from(*amount);
+            coin_dto.serialize(serializer)
+        })
+}
+
+impl<C, BaseG, QuoteC, QuoteG> From<Price<C, QuoteC>> for BasePrice<BaseG, QuoteC, QuoteG>
+where
+    C: Currency,
+    BaseG: Group,
+    QuoteC: Currency,
+    QuoteG: Group,
+{
+    fn from(price: Price<C, QuoteC>) -> Self {
+        Self::new_unchecked(price.amount.into(), price.amount_quote)
+    }
+}
+
+impl<C, BaseG, QuoteC, QuoteG> TryFrom<&BasePrice<BaseG, QuoteC, QuoteG>> for Price<C, QuoteC>
+where
+    C: Currency,
+    BaseG: Group,
+    QuoteC: Currency,
+    QuoteG: Group,
+{
+    type Error = Error;
+
+    fn try_from(value: &BasePrice<BaseG, QuoteC, QuoteG>) -> Result<Self, Self::Error> {
```

Comment:
please refrain from using names generated from the IDE tools
often, they are too generic
what about `value` -> `base`